ThePhasesOfCraftship/1_code_first/assignment/begin/backend/src/index.ts

```diff
@@ -0,0 +1,186 @@
+import express, { Request, Response, NextFunction } from 'express'
+import { PrismaClient, Prisma } from '@prisma/client'
+
+
+const prisma = new PrismaClient()
+// console.log(prisma);
+
+type User = {
+    email: string
+    username: string
+    firstName: string
+    lastName: string
+    password: string
+}
+
+type UserResponse = Omit<User, 'password'>
+
+type ResponseObject = {
+    error: ErrorName | undefined
+    data: UserResponse | undefined
+    success: boolean
+}
+
+const ErrorName = {    
+    UsernameAlreadyTaken: 'UserNameAlreadyTaken',
+    EmailAlreadyInUse: 'EmailAlreadyInUse',
+    ValidationError: 'ValidationError',
+    ServerError: 'ServerError',
+    ClientError: 'ClientError',
+    UserNotFound: 'UserNotFound',      
+}
+
+type ErrorName = typeof ErrorName[keyof typeof ErrorName];
+
+const responseObjectFactory = (error: ErrorName | undefined, data: UserResponse | undefined, success: boolean): ResponseObject => {
+    return {
+        error,
+        data,
+        success
+    }
+}
+
+const basicPasswordGenerator = (length: number): string => {
+    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*()_+-=[]{}|;:,.<>?';
+    let result = '';
+    for (let i = 0; i < length; i++) {
+      result += characters.charAt(Math.floor(Math.random() * characters.length));
+    }
+    return result;
+  }
+
+const createUser = async ({ email, firstName, lastName, username }: Omit<User, 'password'>) => {
+    const user = await prisma.user.create({
+        data: {
+            email,
+            username,
+            firstName,
+            lastName,
+            password: basicPasswordGenerator(10)
+        }
+    })
+    return user
+}
+
+const removePasswordForResponse = (user:  User): UserResponse => {
+    const { password, ...userResponse } = user
+    return userResponse
+}
+
+const getUserByEmail = async (email: string) => {    
+    const user = await prisma.user.findUnique({
+        where: {
+            email
+        }
+    })
+    return user
+
+}
+
+
+const updateUser = async (userId: number, { email, firstName, lastName, username }: Omit<User, 'password'>) => {
+    const user = await prisma.user.update({
+        where: {
+            id: userId
+        },
+        data: {
+            email,
+            firstName,
+            lastName,
+            username
+        }
+    })
+    return user
+}
+
+const app = express()
+const port = 8090
+
+app.use(express.json())
+
+app.listen(port, () => {
+    console.log(`Server is running on port ${port}`)
+})
+
+app.get('/', (req: Request, res: Response) => {
```

Comment:
To keep things clean and minimal you can remove what's not needed and part of the specifications.